test-app/runtime/src/main/cpp/console/Console.cpp

```diff
@@ -0,0 +1,332 @@
+    bindFunctionProperty(context, console, "log", logCallback);
+    bindFunctionProperty(context, console, "warn", warnCallback);
+    bindFunctionProperty(context, console, "dir", dirCallback);
+    bindFunctionProperty(context, console, "trace", traceCallback);
+    bindFunctionProperty(context, console, "time", timeCallback);
+    bindFunctionProperty(context, console, "timeEnd", timeEndCallback);
+
+    return console;
+}
+
+void Console::sendToADBLogcat(const std::string& message, android_LogPriority logPriority) {
+    // split strings into chunks of 4000 characters
+    // __android_log_write can't send more than 4000 to the stdout at a time
+
+    auto messageLength = message.length();
+    int maxStringLength = 4000;
+
+    if (messageLength < maxStringLength) {
+        __android_log_write(logPriority, Console::LOG_TAG, message.c_str());
+    } else {
+        for (int i = 0; i < messageLength; i += maxStringLength) {
+            auto messagePart = message.substr(i, maxStringLength);
+
+            __android_log_write(logPriority, Console::LOG_TAG, messagePart.c_str());
+        }
+    }
+}
+
+void Console::sendToDevToolsFrontEnd(v8::Isolate* isolate, const std::string& message, const std::string& logLevel) {
+    if (!JsV8InspectorClient::inspectorIsConnected()) {
+        return;
+    }
+
+    auto stack = v8::StackTrace::CurrentStackTrace(isolate, 1, v8::StackTrace::StackTraceOptions::kDetailed);
+
+    auto frame = stack->GetFrame(0);
+
+    // will be no-op in non-debuggable builds
+    v8_inspector::V8LogAgentImpl::EntryAdded(message, logLevel, ArgConverter::ConvertToString(frame->GetScriptNameOrSourceURL()), frame->GetLineNumber());
+}
+
+const std::string& buildLogString(const v8::FunctionCallbackInfo<v8::Value>& info) {
+    auto isolate = info.GetIsolate();
+
+    v8::HandleScope scope(isolate);
+
+    std::stringstream ss;
+
+    auto argLen = info.Length();
+    if (argLen) {
+        for (int i = 0; i < argLen; i++) {
+            v8::Local<v8::String> argString;
+            if (info[i]->IsFunction()) {
+                info[i]->ToDetailString(isolate->GetCurrentContext()).ToLocal(&argString);
+            } else if (info[i]->IsObject()) {
+                v8::Local<v8::Object> obj = info[i].As<v8::Object>();
+                argString = JsonStringifyObject(isolate, obj);
+            } else {
+                info[i]->ToDetailString(isolate->GetCurrentContext()).ToLocal(&argString);
+            }
+
+            // separate args with a space
+            if (i != 0) {
+                ss << " ";
+            }
+
+            ss << ArgConverter::ConvertToString(argString);
+        }
+    }
+
+    return ss.str();
+}
+
+void Console::assertCallback(const v8::FunctionCallbackInfo<v8::Value>& info) {
+    auto isolate = info.GetIsolate();
+
+    auto argLen = info.Length();
+    auto expressionPasses = argLen && info[0]->BooleanValue();
+
+    if (!expressionPasses) {
+        std::stringstream ss;
```

Comment:
rename to `assertionError` or something

Reply:
the `expressionPasses` is the first argument which is the test, and determines whether to print an error or not.